### ISSUE TITLE
port-forwarding without any ssh nodes?

### ISSUE THREAD

**Issue by timothy-spencer** [gravitational/teleport#6504]
### What
We would like to be able to port forward and/or ssh into specified kubernetes services.

### How
Ideally, we would be able to specify an ssh application or a ssh proxy endpoint that teleport would let us ssh to without it having to be running teleport.  We could then put something like this in our .ssh/config file:
```
Host gitlab
    Hostname gitlab-gitlab-shell.gitlab
    ProxyJump teleport.<domain>:443
```
And then it seems like we could do `git clone git@gitlab:root/repo.git` and have teleport proxy port 22 in to the gitlab-gitlab-shell service in the gitlab namespace so we could do our git work.

Alternatively, we could probably also work with us being able to use tsh to set up local port-forwarding into this service so we could do a `git clone git://localhost:port/root/repo.git/` to some port that we set up with tsh, though this would be less pleasant because we'd have a long-lived connection that we'd have to nail up at the start of the day and manage as if it were a VPN.

### Why
We would like to be able to run gitlab in a kubernetes cluster and use teleport to manage access.  That all works, and teleport lets us get access to the web UI just fine, but there seems to be no way to tell teleport to proxy in ssh to the `gitlab-gitlab-shell.gitlab` service that was created in k8s so that we can do git pull/clone/etc.  I don't see any good ways of getting the gitlab stuff to also run a teleport node without a lot of hacking because it is managed by helm.

### Workaround
I don't know of any workarounds. But I would be happy to be proven wrong!

**Comment by programmerq** [comment]
@timothy-spencer You can achieve git-through-teleport by setting up your openssh client to work with teleport, and then set up a ProxyJump/ProxyCommand to send your gitlab port 22 ssh connections through a teleport node that has network access to gitlab.

I did a quick test here:

```~/.ssh/config

Host cluster.teleport.cf
Port 3023
ProxyCommand=none

Host node123
Port 3022
StrictHostKeyChecking=yes
ProxyJump %r@proxy.example.com:3023

Host git.example.com
ProxyJump ubuntu@node123
```

That means that when git tries to connect to my `git.example.com` service, it will ProxyJump to my "node123" teleport node. Since I have a `Host` section for that node, that will ProxyJump through my teleport cluster. This is all 100% behind the scenes. There is a measurable time difference, but it isn't significant in my test setup.

Since you have kubernetes in the mix, you could potentially use the kubernetes integration with teleport and then access the kubernetes service via `kubectl port-forward` and get the localhost behavior option. Personally, I'd prefer setting up the double ProxyJump setup described above since it provides a nicer overall user experience. When your teleport session expires, you'll get "could not read from remote repository" errors from git. Once you rerun "tsh login" the git commands continue working as expected.

**Comment by timothy-spencer** [comment]
Yeah, that's the problem.  I have no teleport nodes in my kubernetes cluster.  I only have gitlab running in containers, and it would be very difficult to get teleport going on those gitlab containers because that's all supplied from the gitlab folks upstream.  I could get a teleport node going in a container in the cluster, but it seems like you need teleport running on the pod that we want to talk to for this to work (the gitlab-shell pod in my case).  Please set me straight if I am wrong, though!

I have this working with kubectl port-forward right now, and that works, but it requires us to set up a long-lived connection and use different tools to access resources in the cluster (teleport gui for web access, kubectl port-forward for ssh access).  I'd rather just use teleport for everything, especially since it would be great if we could keep our k8s API hidden from most of our users so that we have less attack surface.

**Comment by programmerq** [comment]
I think I see what you're saying. You can indeed run an ssh_service teleport instance inside a pod that you could use to connect to this endpoint.

While searching for a way to `kubectl port-forward` to stdio to work with an ssh ProxyCommand, I came across this gist that sets up a port-forward and a socat instance to do it for you: https://gist.github.com/sdkks/5e58ea81a719d67f0103d7713b0c4b55

It just fires up a long-running `kubectl port-forward` in the background if it isn't running, and then uses `socat` in the ProxyCommand to send the connection over that forwarded port.

Yet another option, if you have socat in a pod that has access to the endpoint, and the ability to launch it via `kubectl exec`:

```
ProxyCommand kubectl exec -i pod/socatwithaccesstogitlab -- socat tcp-connect: gitlab-gitlab-shell.gitlab:22,reuseaddr,fork stdio
```

You can pass in `--kubeconfig` and `--context` arguments to kubeconfig to force it to use the correct config/context regardless of what your default context is. For a teleport protected kubernetes setup, your workflow would be to `teleport login` and then your git commands would get passed over that.

**Comment by timothy-spencer** [comment]
Thanks for looking at this!

So we'd love it if we could avoid using kubectl, because our users are not kubernetes users, they are gitlab users.  I'd rather not expose the kubernetes API to them if we can help it.  Our ideal situation would be if we could use tsh to port forward to a gitlab service so that we could use it as a ProxyCommand pretty much exactly like you are doing with kubectl/socat.  In our particular case, I'd even be fine with just being able to ProxyJump into one particular service, though that would not be as generally useful as being able to use tsh to port forward in.

This would be superior because we wouldn't need to expose the k8s API to users and because presumably we could set tsh up to only allow certain users to connect to particular hosts/services/ports instead of being able to connect to anything.  We really like the FedRAMP-friendly auditing and control stuff with your product, so being able to manage who can connect to what with just this one tool would really give us a great compliance story.

**Comment by timothy-spencer** [comment]
As an example, something like `tssh -L 2222:gitlab-gitlab-shell.gitlab:22 --port-forward-only --local git clone git@gitlab:root/repo.git` would be super cool.  In the teleport.yml file, we would be able to specify what hosts/ports we would allow authorized teleport users to be able to connect to, just like we do with apps and so on.

Is that architecturally difficult?  It seems like this would be useful.  It would certainly be useful for us.  :-)  Thanks, and have fun!

**Comment by programmerq** [comment]
Fixed by https://github.com/gravitational/teleport/pull/13455
